fastparquet/api.py

```diff
@@ -257,6 +256,13 @@ def statistics(self):
             self._statistics = statistics(self)
         return self._statistics
 
+    @property
+    def key_value_metadata(self):
+        if self._kvm is None:
+            self._kvm = {k.key.decode(): k.value.decode()
```

Comment:
Could there be a case where this has already been decoded?

Reply:
I am not sure to understand the question. The only time when this gets decoded is when this property is called.

Reply:
I can add `if` in case it is. It is not in the test cases.
But in the other branch where this could be set by the user (custom metadata), it's true that was encoding the content of `ParquetFile.fmd.key_value_metadata` so as to 'comply' with having them always encoded.

Reply:
> encoding the content of ParquetFile.fmd.key_value_metadata so as to 'comply' with having them always encoded.

That's probably good enough...

Reply:
@martindurant I am sorry, I don't understand if I should do something more then?

Hmmm,
I think we could provide `fastparquet` users willing to modify custom metadata with a _bare_ function to ensure this encoding is done. So this would be an extract of the previous PR (custome metadata), stripped from file modifications. I could add it in `util.py`. Something like:

```python
def update_custom_metadata(fmd: ThriftObject, custom_metadata : dict):
       """Update custom metadata stored in thrift object.
        Update strategy depend if key found in new custom metadata is also
        found in already existing custom metadata within thrift object, as well
        as its value.
        
          - If not found in existing, it is added.
          - If found in existing, it is updated.
          - If its value is `None`, it is not added, and if found in existing,
            it is removed from existing.

        Parameters
        ----------
        fmd : ThriftObject
            Thrift object which metadata is to update.
        custom_metadata : dict
            Key-value metadata to update in thrift object.
        """
        kvm = fmd.key_value_metadata
        # Spare list of keys.
        kvm_keys = [item.key for item in kvm]
        for key, value in custom_metadata.items():
            key_b = key.encode()
            if key_b in kvm_keys:
                idx = kvm_keys.index(key_b)
                if value is None:
                    # Remove item.
                    del kvm[idx]
                    # Update 'kvm_keys' as well, for keeping indexing
                    # up-to-date.
                    del kvm_keys[idx]
                else:
                    # Replace item.
                    kvm[idx] = parquet_thrift.KeyValue(key=key_b, value=value.encode())
            elif value is not None:
                kvm.append(parquet_thrift.KeyValue(key=key_b, value=value.encode()))
        fmd.key_value_metadata = kvm
```

The user would then be free to apply the way it is needed.
Do you think this would be a relevant addition?

Reply:
Yes, that function is exactly what I was getting at, and sketched out previously. It should be a separate PR, though. Also, it would make sense to optionally load from and write to a file (_metadata or otherwise). Writing to a non-_metadata file only makes sense where you can open a file for update ("rb+"), which is probably only local files.